src/Features/CSharp/Portable/ImplementInterface/AbstractChangeImplementionCodeRefactoringProvider.cs

```diff
@@ -0,0 +1,267 @@
+﻿// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
+
+#nullable enable
+
+using System;
+using System.Collections.Generic;
+using System.Collections.Immutable;
+using System.Linq;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.CodeAnalysis;
+using Microsoft.CodeAnalysis.CodeActions;
+using Microsoft.CodeAnalysis.CodeRefactorings;
+using Microsoft.CodeAnalysis.CSharp.Syntax;
+using Microsoft.CodeAnalysis.Editing;
+using Microsoft.CodeAnalysis.PooledObjects;
+using Microsoft.CodeAnalysis.Shared.Extensions;
+using Microsoft.CodeAnalysis.Text;
+using Roslyn.Utilities;
+
+namespace Microsoft.CodeAnalysis.CSharp.ImplementInterface
+{
+    // A mapping from a concrete member to the interface members it implements.
+    using MemberImplementationMap = OrderedMultiDictionary<ISymbol, ISymbol>;
+
+    internal abstract class AbstractChangeImplementionCodeRefactoringProvider : CodeRefactoringProvider
+    {
+        private static readonly SymbolDisplayFormat NameAndTypeParametersFormat =
+            new SymbolDisplayFormat(
+                globalNamespaceStyle: SymbolDisplayGlobalNamespaceStyle.Omitted,
+                typeQualificationStyle: SymbolDisplayTypeQualificationStyle.NameOnly,
+                genericsOptions: SymbolDisplayGenericsOptions.IncludeTypeParameters,
+                miscellaneousOptions: SymbolDisplayMiscellaneousOptions.UseSpecialTypes);
+
+        protected abstract string Implement_0 { get; }
+        protected abstract string Implement_all_interfaces { get; }
+        protected abstract string Implement { get; }
+
+        protected abstract bool CheckExplicitName(ExplicitInterfaceSpecifierSyntax? explicitName);
+        protected abstract bool CheckMember(ISymbol member);
+        protected abstract SyntaxNode ChangeImplementation(SyntaxGenerator generator, SyntaxNode currentDecl, ISymbol interfaceMember);
+        protected abstract Task UpdateReferencesAsync(Project project, Dictionary<Document, SyntaxEditor> documentToEditor, ISymbol implMember, INamedTypeSymbol containingType, CancellationToken cancellationToken);
+
+        public sealed override async Task ComputeRefactoringsAsync(CodeRefactoringContext context)
+        {
+            var (container, explicitName, name) = await GetContainerAsync(context).ConfigureAwait(false);
+            if (container == null)
+                return;
+
+            if (!CheckExplicitName(explicitName))
+                return;
+
+            var (document, _, cancellationToken) = context;
+            var semanticModel = await document.GetSemanticModelAsync(cancellationToken).ConfigureAwait(false);
+            var member = semanticModel.GetDeclaredSymbol(container, cancellationToken) ??
+                throw new InvalidOperationException();
```

Comment:
Nit: add some string to the message here in case we ever have to debug some edge case in the wild. Doesn't have to be localized. Alternatively consider Contract.ThrowIfNull().

Reply:
sure